File: clients/java/src/test/java/io/volant/MetadataTopicsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

/** v0.116: Java Client Metadata topic filter ({@code metadata(List)}). */
class MetadataTopicsTest {
    private static final int TIMEOUT = 7;

    @Test
    void metadataSendsEmptyTopicsList() throws Exception {
        try (MetaTopicsStub srv = MetaTopicsStub.start()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                Metadata got = c.metadata();
                assertTrue(got.brokers.isEmpty());
                assertTrue(got.topics.isEmpty());
            }
            assertEquals(1, srv.metadataCount.get());
            assertEquals(List.of(List.of()), srv.seenTopics());
        }
    }

    @Test
    void metadataListEncodesNamedFilter() throws Exception {
        try (MetaTopicsStub srv = MetaTopicsStub.start()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                Metadata got = c.metadata(List.of("events"));
                assertTrue(got.brokers.isEmpty());
            }
            assertEquals(1, srv.metadataCount.get());
            assertEquals(List.of(List.of("events")), srv.seenTopics());
        }
    }

    @Test
    void metadataEmptyListMatchesAllTopics() throws Exception {
        try (MetaTopicsStub srv = MetaTopicsStub.start()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.metadata();
                c.metadata(Collections.emptyList());
                c.metadata(List.of());
            }
            assertEquals(3, srv.metadataCount.get());
            List<List<String>> seen = srv.seenTopics();
            assertEquals(3, seen.size());
            for (List<String> topics : seen) {
                assertTrue(topics.isEmpty());
            }
        }
    }

    @Test
    void metadataStillRetriesTimeout() throws Exception {
        try (MetaTopicsStub srv = MetaTopicsStub.start()) {
            srv.metadataCodes.add(TIMEOUT);
            srv.metadataCodes.add(0);
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.setMaxRetries(2);
                c.setRetryBackoffMs(0);
                Metadata got = c.metadata();
                assertTrue(got.brokers.isEmpty());
                assertTrue(got.topics.isEmpty());
            }
            assertEquals(2, srv.metadataCount.get());
            assertEquals(List.of(List.of(), List.of()), srv.seenTopics());
        }
    }

    private static final class MetaTopicsStub implements AutoCloseable {
        final int port;
        final List<Integer> metadataCodes = new CopyOnWriteArrayList<>();
        final AtomicInteger metadataCount = new AtomicInteger();
        private final List<List<String>> seenTopics = new CopyOnWriteArrayList<>();
        private final ServerSocket listen;
        private final Thread acceptThread;

        static MetaTopicsStub start() throws IOException {
            return new MetaTopicsStub();
        }

        private MetaTopicsStub() throws IOException {
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            acceptThread = new Thread(
                    () -> {
                        while (!listen.isClosed()) {
                            try {
                                Socket conn = listen.accept();
                                Thread t = new Thread(() -> serve(conn), "volant-meta-topics-conn");
                                t.setDaemon(true);
                                t.start();
                            } catch (IOException e) {
                                return;
                            }
                        }
                    },
                    "volant-meta-topics-accept");
            acceptThread.setDaemon(true);
            acceptThread.start();
        }

        List<List<String>> seenTopics() {
            return new ArrayList<>(seenTopics);
        }

        private void serve(Socket conn) {
            try {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    if (d.frame.opcode != Codec.OP_METADATA) {
                        throw new ProtocolException("unexpected opcode " + d.frame.opcode);
                    }
                    Codec.MetadataRequest req = Codec.decodeMetadataRequest(d.frame.payload);
                    metadataCount.incrementAndGet();
                    seenTopics.add(new ArrayList<>(req.topics));
                    int code = 0;
                    if (!metadataCodes.isEmpty()) {
                        code = metadataCodes.remove(0);
                    }
                    byte[] payload;
                    int replyOp = Codec.OP_METADATA;
                    if (code != 0) {
                        replyOp = Codec.OP_ERROR;
                        payload = Codec.encodeErrorResponse(new Codec.ErrorResponse(code, ""));
                    } else {
                        payload = Codec.encodeMetadataResponse(
                                new Metadata(Collections.emptyList(), Collections.emptyList()));
                    }
                    out.write(Frame.encode(replyOp, d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception ignored) {
                // client closed or timeout
            } finally {
                try {
                    conn.close();
                } catch (IOException ignored) {
                    // best-effort
                }
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                acceptThread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/go/metadata_topics_test.go
package volant_test

import (
	"net"
	"sync"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

// v0.116: Go Client Metadata topic filter (MetadataTopics).

type metaTopicsStub struct {
	mu         sync.Mutex
	codes      []uint16
	seenTopics [][]string
	count      int
	ln         net.Listener
}

func startMetaTopicsStub(t *testing.T, codes []uint16) (addr string, stub *metaTopicsStub, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	stub = &metaTopicsStub{codes: append([]uint16(nil), codes...), ln: ln}
	done := make(chan struct{})
	go func() {
		defer close(done)
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			go stub.serve(conn)
		}
	}()
	return ln.Addr().String(), stub, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func (s *metaTopicsStub) serve(conn net.Conn) {
	defer conn.Close()
	_ = conn.SetDeadline(time.Now().Add(10 * time.Second))
	buf := []byte{}
	tmp := make([]byte, 4096)
	for {
		f, rest, err := frame.TryDecode(buf)
		if err != nil {
			return
		}
		if f == nil {
			n, err := conn.Read(tmp)
			if n > 0 {
				buf = append(buf, tmp[:n]...)
			}
			if err != nil {
				return
			}
			continue
		}
		buf = append([]byte(nil), rest...)
		raw, err := s.handle(f)
		if err != nil {
			return
		}
		if _, err := conn.Write(raw); err != nil {
			return
		}
		_ = conn.SetDeadline(time.Now().Add(10 * time.Second))
	}
}

func (s *metaTopicsStub) handle(f *frame.Frame) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if f.Opcode != codec.OpMetadata {
		return nil, &frame.ProtocolError{Msg: "unexpected opcode"}
	}
	req, err := codec.DecodeMetadataRequest(f.Payload)
	if err != nil {
		return nil, err
	}
	s.count++
	topics := append([]string(nil), req.Topics...)
	s.seenTopics = append(s.seenTopics, topics)
	code := uint16(0)
	if len(s.codes) > 0 {
		code = s.codes[0]
		s.codes = s.codes[1:]
	}
	var payload []byte
	replyOp := uint16(codec.OpMetadata)
	if code != 0 {
		payload, err = codec.EncodeErrorResponse(codec.ErrorResponse{Code: code})
		replyOp = codec.OpError
	} else {
		payload, err = codec.EncodeMetadataResponse(codec.MetadataResponse{})
	}
	if err != nil {
		return nil, err
	}
	return frame.Encode(replyOp, f.CorrelationID, payload)
}

func (s *metaTopicsStub) rpcs() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.count
}

func (s *metaTopicsStub) seen() [][]string {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([][]string, len(s.seenTopics))
	for i, t := range s.seenTopics {
		out[i] = append([]string(nil), t...)
	}
	return out
}

func TestMetadataSendsEmptyTopicsList(t *testing.T) {
	addr, stub, stop := startMetaTopicsStub(t, nil)
	defer stop()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	got, err := c.Metadata()
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Brokers) != 0 || len(got.Topics) != 0 {
		t.Fatalf("metadata %v want empty", got)
	}
	if n := stub.rpcs(); n != 1 {
		t.Fatalf("metadata count %d want 1", n)
	}
	seen := stub.seen()
	if len(seen) != 1 || len(seen[0]) != 0 {
		t.Fatalf("seen topics %v want [[]]", seen)
	}
}

func TestMetadataTopicsEncodesNamedFilter(t *testing.T) {
	addr, stub, stop := startMetaTopicsStub(t, nil)
	defer stop()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	got, err := c.MetadataTopics([]string{"events"})
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Brokers) != 0 {
		t.Fatalf("brokers %v want empty", got.Brokers)
	}
	if n := stub.rpcs(); n != 1 {
		t.Fatalf("metadata count %d want 1", n)
	}
	seen := stub.seen()
	if len(seen) != 1 || len(seen[0]) != 1 || seen[0][0] != "events" {
		t.Fatalf("seen topics %v want [[events]]", seen)
	}
}

func TestMetadataTopicsEmptyMatchesMetadata(t *testing.T) {
	addr, stub, stop := startMetaTopicsStub(t, nil)
	defer stop()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if _, err := c.Metadata(); err != nil {
		t.Fatal(err)
	}
	if _, err := c.MetadataTopics(nil); err != nil {
		t.Fatal(err)
	}
	if _, err := c.MetadataTopics([]string{}); err != nil {
		t.Fatal(err)
	}
	if n := stub.rpcs(); n != 3 {
		t.Fatalf("metadata count %d want 3", n)
	}
	seen := stub.seen()
	if len(seen) != 3 {
		t.Fatalf("seen %v", seen)
	}
	for i, topics := range seen {
		if len(topics) != 0 {
			t.Fatalf("seen[%d]=%v want empty", i, topics)
		}
	}
}

func TestMetadataStillRetriesTimeout(t *testing.T) {
	addr, stub, stop := startMetaTopicsStub(t, []uint16{timeoutCode, 0})
	defer stop()

	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	c.SetMaxRetries(2)
	c.SetRetryBackoff(0)

	got, err := c.Metadata()
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Brokers) != 0 || len(got.Topics) != 0 {
		t.Fatalf("metadata %v want empty", got)
	}
	if n := stub.rpcs(); n != 2 {
		t.Fatalf("metadata count %d want 2", n)
	}
	seen := stub.seen()
	if len(seen) != 2 || len(seen[0]) != 0 || len(seen[1]) != 0 {
		t.Fatalf("seen topics %v want two empty lists", seen)
	}
}
